### PR TITLE
Card Tricks: Adapt to current concept tree progression #1990

### LINKED ISSUE #1990: Card Tricks: Adapt to current concept tree progression
## Getting Started

If you have not yet contributed to concept exercises before, this task requires some upfront reading to acquire the necessary background knowledge.

Here you can read about what Concept Exercises are and how they are structured:

- [What are Concept Exercises?](https://exercism.org/docs/building/product/concept-exercises)
- [Concept Exercises](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concepts](https://exercism.org/docs/building/tracks/concepts)

## Motivation

Currently the concept exercise "Card Tricks" has some issues.
1. It requires the student to know about `for-loops` which are taught later in the [syllabus](https://exercism.org/tracks/go/concepts) and depend on `slices`.
2. It requires the student to know about variadic functions (`append` in this case) which were not introduced yet.
3. `multiple-return-values` is now part of the `functions` concept and should not be taught as part of Card Tricks anymore.

These issues should be addressed.

## Tasks

- [ ] Remove the task about `PrefilledSlice` from all the relevant exercise files (it requires for loops).
- [ ] Optional (can also be done later as a separate issue/PR): As replacement for `PrefilledSlice` it would be good to have another task that requires the student to create a slice. Additionally, the student should get a task that requires to define a variadic function.
- [ ] Change `GetItem` to have only one return value. In case the index is out of range, `-1` should be returned. Make sure to apply the change consistently (check hints, instructions etc.).
- [ ] Add the content of `concepts/variadic-functions/introductions.md` to the introductions file of the exercise. A good structure would be
  - `# Introduction`
  - `## Slices` (the content in slices should mention that append is a variadic function and refer to the details below)
  - `## Variadic Functions`
- [ ] Update the root level config.json file:
  - Remove `multiple-return-values` from the concepts taught by `card-tricks`.
  - Add `variadic-functions` to the concepts taught by `card-tricks`.
  - There are 2 other exercises that have `multiple-return-values` as prerequisite. Change the concept name to `functions` for those.
  
## Out of Scope

- Writing the `variadic-functions` concept is a separate issue.
- Entirely removing the `multiple-return-values` concept will be separate issue.

---

If you want to work on this issue, just leave a comment below and go ahead. No need to wait for a reply.

### PR DESCRIPTION
**Pull request by norbs57**
Adapted card tricks to  current concept tree progression 
Fixes #1990

* Removed the task about `PrefilledSlice` from all the relevant exercise files
* Added function `FavoriteCards` which returns a new slice filled with numbers 2,5,9. Added to docs and added a test.
* Changed `GetItem` to have only one return value. Updated tests accordingly.
* Added code in `TestSetItem` which checks that `SetItem(slice,index)` modifies the
  existing slice as long as `index` is within slice bounds.
* Added a variadic function `PrependItems`.  Added to docs and added a test.
* Added the content of `concepts/variadic-functions/introductions.md` to the introduction.
* Added more info about the inbuilt `append` function to hints and instructions.
* Updated the root level `config.json` in line with #1990 recommendations

### CODE REVIEW

File: exercises/concept/card-tricks/card_tricks_test.go [hunk 1]
```diff
@@ -133,121 +132,141 @@ func TestSetItem(t *testing.T) {
...
 					tt.args.slice, tt.args.index, tt.args.value, got, tt.want)
 			}
+			if len(tt.args.slice) == len(got) {
+				for i := range got {
+					got[i] = -1
+				}
+				if !slicesEqual(got, tt.args.slice) {
+					t.Errorf("SetItem(slice:%v, index:%v) does not return the modified input slice)", tt.args.slice,
+						tt.args.value)
+				}
```
Comment (junedev) [inline exercises/concept/card-tricks/card_tricks_test.go:147]: since we already established the values are equal, I would recommend something like this to check the reference is still the same: https://go.dev/play/p/oHmY2lUpO5L

File: exercises/concept/card-tricks/card_tricks_test.go [hunk 2]
```diff
@@ -133,121 +132,141 @@ func TestSetItem(t *testing.T) {
...
-			name: "Remove the first item",
+			name: "prepend zero items",
 			args: args{
-				slice: []int{3, 4, 5, 6},
-				index: 0,
+				slice: []int{5, 2, 10, 6, 8, 7, 0, 9},
+				value: []int{},
 			},
-			want: []int{4, 5, 6},
+			want: []int{5, 2, 10, 6, 8, 7, 0, 9},
```
Comment (junedev) [inline exercises/concept/card-tricks/card_tricks_test.go:237]: I would recommend to move these edge cases to the end of the test list so the student sees a usual example in the test case first.

**Comment by junedev** [comment]
Looks good overall! Thanks a lot. I left some ideas for minor improvements.

**Comment by junedev** [comment]
Sidenote: When you mention the issue in the PR description with `Fixes #1234` than GitHub will automatically link the PR to the issue and close the issue once the PR is merged. See [GitHub docs](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more details.

File: exercises/concept/card-tricks/.docs/introduction.md [hunk 3]
```diff
@@ -28,27 +28,88 @@ If you don't specify an ending index, it defaults to the length of the slice.
...
 
-## Multiple Return Values
+Go provides a built-in `append` function to append new elements to a slice: 
 
-Go functions and methods can return multiple values.
-Very often, a second return value is used to return an error.
-For example:
+```go 
+a := []int{1,3}
+b := append(a,4,2)  // b == []int{1,3,4,2}
```
[the remark this answers is not in the document]
Comment (norbs57) [reply exercises/concept/card-tricks/.docs/introduction.md:35]: Done

File: exercises/concept/card-tricks/.docs/hints.md [hunk 4]
```diff
@@ -2,29 +2,47 @@
...
+b := []int{4,2,6}
+c := append(a,b...)  
+fmt.Println(c)
+// output is [1 3 4 2 6]
+```
+
+## 1. AllItems 
+
+- There are several ways to create a new slice. The simplest is `var s []int`.
+- You get more options with the [`make`][make-builtin] function.
```
[the remark this answers is not in the document]
Comment (norbs57) [reply exercises/concept/card-tricks/.docs/hints.md:20]: I have removed these. They were a leftover from a previous version of the exercise.

File: exercises/concept/card-tricks/.docs/hints.md [hunk 5]
```diff
@@ -2,29 +2,47 @@
...
-- There are several ways to create a new slice. The simplest is `var s []int`. More options you get with the [`make`][make-builtin] function.
-- Use the techniques from 2. to fill the slice with the correct values in a `for` loop.
+- Removing an item from a slice can be done by appending the part after `index`
+  to the part before index. 
 
-## 4. Remove a card from the stack
+## 5. Prepend cards to the stack
 
-- Removing an item from a slice can be done by appending the part after `index` to the part before index.
+ - Adding items to the front of a slice can be done by appending the elements of
```
[the remark this answers is not in the document]
Comment (norbs57) [reply exercises/concept/card-tricks/.docs/hints.md:41]: Done

File: exercises/concept/card-tricks/.docs/introduction.md [hunk 6]
```diff
@@ -28,27 +28,88 @@ If you don't specify an ending index, it defaults to the length of the slice.
 newSlice := withData[2:4] // newSlice == []int{2,3}
 ```
 
-## Multiple Return Values
+Go provides a built-in `append` function to append new elements to a slice:
```
[the remark this answers is not in the document]
Comment (norbs57) [reply exercises/concept/card-tricks/.docs/introduction.md:31]: I have replaced this with the phrase below from the official [Go Tour](https://go.dev/tour/moretypes/15). It is also quite similar to the formulation in [Go slices](https://go.dev/blog/slices-intro).  
```text
It is common to append new elements to a slice, and so Go provides a built-in `append` function.
```

File: exercises/concept/card-tricks/.docs/instructions.md [hunk 7]
```diff
@@ -4,70 +4,98 @@ As a magician-to-be, Elyse needs to practice some basics. She has a stack of car
...
 ```go
-RemoveItem([]int{3, 2, 6, 4, 8}, 2)
-// Output: []int{3, 2, 4, 8}
+cards := RemoveItem([]int{3, 2, 6, 4, 8}, 11)
+fmt.Println(cards) 
+// Output: [3 2 6 4 8]
 ```
 
-If the index is out of bounds (ie. if it is negative or after the end of the stack), we want to leave the stack unchanged:
+## 5. PrependItems
```
[the remark this answers is not in the document]
Comment (norbs57) [reply exercises/concept/card-tricks/.docs/instructions.md:74]: Done

File: exercises/concept/card-tricks/.docs/instructions.md [hunk 8]
```diff
@@ -4,70 +4,98 @@ As a magician-to-be, Elyse needs to practice some basics. She has a stack of car
 
 To make things a bit easier she only uses the cards 1 to 10.
 
-## 1. Retrieve a card from a stack
+## 1. Create a slice with all the cards 
+
+Return a slice with all cards from 1 to 10
```
[the remark this answers is not in the document]
Comment (norbs57) [reply exercises/concept/card-tricks/.docs/instructions.md:9]: I like that suggestion - I have replaced function `AllItems` with `FavoriteCards` accordingly.

File: exercises/concept/card-tricks/.docs/instructions.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (norbs57) [reply exercises/concept/card-tricks/.docs/instructions.md:74]: Done

File: exercises/concept/card-tricks/card_tricks_test.go (hunk 1 above)
Comment (norbs57) [reply exercises/concept/card-tricks/card_tricks_test.go:147]: Nice! Done.

File: exercises/concept/card-tricks/card_tricks_test.go (hunk 2 above)
Comment (norbs57) [reply exercises/concept/card-tricks/card_tricks_test.go:237]: Done.

File: exercises/concept/card-tricks/.docs/instructions.md [hunk 9]
```diff
@@ -4,70 +4,98 @@ As a magician-to-be, Elyse needs to practice some basics. She has a stack of car
...
 ```
 
-If the index is out of bounds (ie. if it is negative or after the end of the stack), we want to leave the stack unchanged:
+## 5. PrependItems
+
+Add values at the front of a stack. Note that `PrependItems` is a variadic
+function, its type is: 
+
+```go 
+func PrependItems(slice []int, value ...int) []int
```
[the remark this answers is not in the document]
Comment (norbs57) [reply exercises/concept/card-tricks/.docs/instructions.md:80]: Removed.

File: exercises/concept/card-tricks/.docs/introduction.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (junedev) [reply exercises/concept/card-tricks/.docs/introduction.md]: That's ok. In general though keep in mind that we try to be even more accessible than existing resources. For example in this case imagine a non-native speaker that might struggle with the word "append" for whatever reason (maybe the translations in their native language are ambiguous). For them it would be valuable to read append explained with "add to the end" instead of having "append" explained with "append".

[the remark this answers is not in the document]
Comment (norbs57) [reply exercises/concept/card-tricks/.docs/introduction.md]: Ok, I will use the "add" explanation. I have added a [yourbasic](https://yourbasic.org/golang/append-explained/) link.  I like the explanation of `append` on the web site. Is it an appropriate link with regards to our "linking policies"?

[the remark this answers is not in the document]
Comment (junedev) [reply exercises/concept/card-tricks/.docs/introduction.md]: I don't see a reason why not as long as the content is correct.

File: exercises/concept/card-tricks/.docs/instructions.md [hunk 10]
```diff
@@ -4,70 +4,89 @@ As a magician-to-be, Elyse needs to practice some basics. She has a stack of car
 
 To make things a bit easier she only uses the cards 1 to 10.
 
-## 1. Retrieve a card from a stack
+## 1. Create a slice with certain cards
```
Comment (junedev) [inline exercises/concept/card-tricks/.docs/instructions.md:7]: Headers in this file and in hints need to match, otherwise the website has issues to connect the "show hints" button to the correct content.

Comment (norbs57) [reply exercises/concept/card-tricks/.docs/instructions.md]: Done.

Comment (junedev) [reply exercises/concept/card-tricks/.docs/instructions.md]: The header should be a sentence like "Create a slice with certain cards" as you had it before in instructions, now you changed it to be the function name (like it was in hints).

Comment (norbs57) [reply exercises/concept/card-tricks/.docs/instructions.md:7]: Done.

**Review by junedev — APPROVED** [review]
